### ISSUE TITLE
Ruby 2.5 Support

### ISSUE THREAD

**Issue by headius** [jruby/jruby#4876]
This issue will track Ruby 2.5 support for JRuby. This will be released in JRuby 9.3.0.0 or JRuby 10.

As with previous releases, we will base our checklist on MRI's NEWS file: https://github.com/ruby/ruby/blob/trunk/NEWS
This bug will track all NEWS items new in Ruby 2.5. This list is based off https://github.com/ruby/ruby/blob/trunk/NEWS.

Only changes relevant to JRuby are listed here. MRI-specific internal changes and features are not included.

NOTE: Pull requests should be done against the `ruby-2.5` branch.

### Language changes

- [x] Top-level constant look-up is removed.  [Feature #11547](https://bugs.ruby-lang.org/issues/11547)
- [x] rescue/else/ensure are allowed inside do/end blocks.  [Feature #12906](https://bugs.ruby-lang.org/issues/12906)
- [ ] refinements take place in string interpolations.  [Feature #13812](https://bugs.ruby-lang.org/issues/13812)

#### Array

- [x]  `Array#append`  [Feature #12746](https://bugs.ruby-lang.org/issues/12746)
- [x]  `Array#prepend`  [Feature #12746](https://bugs.ruby-lang.org/issues/12746)

#### Data

- [x] Now deprecated [Feature #3072](https://bugs.ruby-lang.org/issues/3072)

#### Dir

- [x] `Dir.glob` provides new optional keyword argument, `:base`. [Feature #13056](https://bugs.ruby-lang.org/issues/13056)
- [x] `Dir.children`  [Feature #11302](https://bugs.ruby-lang.org/issues/11302)
- [x] `Dir.each_child`  [Feature #11302](https://bugs.ruby-lang.org/issues/11302)

#### File

- [ ]  :newline option to `File.open` implies text mode now.  [Bug #13350](https://bugs.ruby-lang.org/issues/13350)
- [x]  `File#path` now raises an IOError for files opened with `File::Constants::TMPFILE` option. [Feature #13568](https://bugs.ruby-lang.org/issues/13568)
- [x]  Add `File.lutime`  [Feature #4052](https://bugs.ruby-lang.org/issues/4052)

#### Hash

- [x] `Hash#transform_keys`  [Feature #13583](https://bugs.ruby-lang.org/issues/13583)
- [x] `Hash#transform_keys!`  [Feature #13583](https://bugs.ruby-lang.org/issues/13583)
- [x] `Hash#slice`  [Feature #8499](https://bugs.ruby-lang.org/issues/8499)

#### IO

- [ ] `IO#pread`  [Feature #4532](https://bugs.ruby-lang.org/issues/4532)
- [ ] `IO#pwrite`  [Feature #4532](https://bugs.ruby-lang.org/issues/4532)
- [ ] `IO#copy_stream` tries copy offload with `copy_file_range(2)` [Feature #13867](https://bugs.ruby-lang.org/issues/13867)
- [x] `IO#write` accepts multiple arguments  [Feature #9323](https://bugs.ruby-lang.org/issues/9323)

#### IOError

- [x] exception message "stream closed" is changed  [Bug #13405](https://bugs.ruby-lang.org/issues/13405)

#### Integer

- [x] `Integer.sqrt`  [Feature #13219](https://bugs.ruby-lang.org/issues/13219)
- [x] `Integer#step` does no longer rescue exceptions when given a step value which cannot be compared with `#>` to 0. [Feature #7688](https://bugs.ruby-lang.org/issues/7688)
- [x] `Integer#{round,floor,ceil,truncate}` now always return an Integer. [Bug #13420](https://bugs.ruby-lang.org/issues/13420)
- [x] `Integer#pow` added

#### Kernel

- [x] `Kernel#yield_self`  [Feature #6721](https://bugs.ruby-lang.org/issues/6721)
- [x] `Kernel#pp` [Feature #14123](https://bugs.ruby-lang.org/issues/14123)

#### Module

- [x] `Module#attr`, `attr_accessor`, `attr_reader`, `attr_writer` are now public [Feature #14132]
- [x] `Module#define_method`, `alias_method`, `undef_method`, `remove_method` are now public [Feature #14133]

#### Numeric

- [x] Numerical comparison operators (`<` ,`<=`, `>=`, `>`) no longer rescue exceptions of `#coerce`. Return nil in `#coerce` if the coercion is impossible. [Feature #7688](https://bugs.ruby-lang.org/issues/7688)

#### Process

- [ ] Precision of `Process.times` is improved if `getrusage(2)` exists. [Feature #11952](https://bugs.ruby-lang.org/issues/11952)

#### Range

- [x] `Range#initialize` no longer rescue exceptions when comparing begin and
    end with `#<=>` and raise a "bad value for range" ArgumentError
    but instead let the exception from the `#<=>` call go through.
    [Feature #7688](https://bugs.ruby-lang.org/issues/7688)

#### Regexp

- [ ] Update to Onigmo 6.1.1.
- [ ] Support absent operator https://github.com/k-takata/Onigmo/issues/82

#### String

- [x] `String#-@` deduplicates unfrozen strings.  Already-frozen strings remain unchanged for compatibility.  [Feature #13077](https://bugs.ruby-lang.org/issues/13077)
- [x] -"literal" (`String#-@`) optimized to return the same object (same as "literal".freeze in Ruby 2.1+) [Feature #13295](https://bugs.ruby-lang.org/issues/13295)
- [x] `String#{casecmp,casecmp?}` now return nil for non-string arguments instead of raising a TypeError. [Bug #13312](https://bugs.ruby-lang.org/issues/13312)
- [x] `String#delete_prefix` is added to remove prefix [Feature #12694](https://bugs.ruby-lang.org/issues/12694)
- [x] `String#delete_prefix!` is added to remove prefix destructively [Feature #12694](https://bugs.ruby-lang.org/issues/12694)
- [x] `String#delete_suffix` is added to remove suffix [Feature #13665](https://bugs.ruby-lang.org/issues/13665)
- [x] `String#delete_suffix!` is added to remove suffix destructively [Feature #13665](https://bugs.ruby-lang.org/issues/13665)
- [ ] `String#each_grapheme_cluster` and `String#grapheme_clusters` is added to enumerate grapheme clusters [Feature #13780](https://bugs.ruby-lang.org/issues/13780)
- [x] `String#start_with?` supports regexp [Feature #13712](https://bugs.ruby-lang.org/issues/13712)

#### Regexp/String:

- [x] Update Unicode version from 9.0.0 to 10.0.0 [Feature #13685](https://bugs.ruby-lang.org/issues/13685)

#### Thread

- [x] `Thread#fetch`  [Feature #13009](https://bugs.ruby-lang.org/issues/13009) (#5089)
- [x] Description set by `Thread#name=` is now visible on Windows 10.

#### Time

- [x] `Time#at` receives 3rd argument which specifies the unit of 2nd argument. [Feature #13919](https://bugs.ruby-lang.org/issues/13919)

#### KeyError

- [x] `KeyError#receiver` [Feature #12063](https://bugs.ruby-lang.org/issues/12063)
- [x] `KeyError#key`      [Feature #12063](https://bugs.ruby-lang.org/issues/12063)

#### Bundler

- [x] Add Bundler to Standard Library. [Feature #12733](https://bugs.ruby-lang.org/issues/12733)

#### Pathname

- [x] New method: `Pathname#glob` [Feature #7360](https://bugs.ruby-lang.org/issues/7360) (Started in #5095)

#### Psych

- [x] Update to Psych 3.0.0.
- [x] Add `:symbolize_names` option to `Psych.load`, `Psych.safe_load` like `JSON.parse` https://github.com/ruby/psych/pull/333, https://github.com/ruby/psych/pull/337
- [x] Add `Psych::Handler#event_location` https://github.com/ruby/psych/pull/326
- [x] Make `frozen string literal = true` https://github.com/ruby/psych/pull/320
- [x] Preserve time zone offset when deserializing times https://github.com/ruby/psych/pull/316
- [x] Removed deprecated method aliases for `syck` gem https://github.com/ruby/psych/pull/312

#### RbConfig

- [x] New constants: `RbConfig::LIMITS` is added to provide the limits of C types. This is available when rbconfig/sizeof is required. (Partial support implemented, additional types can be added as needed)

#### Ripper

- [x] New method: `Ripper#state` is added to tell the state of scanner. [Feature #13686](https://bugs.ruby-lang.org/issues/13686)
- [x] New constants: `Ripper::EXPR_BEG` and so on for `Ripper#state`.

#### SecureRandom

- [x] New methods: `SecureRandom.alphanumeric`

#### Set

- [x] Add `Set#to_` as alias to `#inspect` [Feature #13676](https://bugs.ruby-lang.org/issues/13676)
- [x] Add `Set#===` as alias to `#include?` [Feature #13801](https://bugs.ruby-lang.org/issues/13801)
- [x] Add `Set#reset` [Feature #6589](https://bugs.ruby-lang.org/issues/6589)

#### StringIO

- [x] `StringIO#write` accepts multiple arguments

#### StringScanner

- [x] Add `StringScanner#size`  [Feature #836](https://bugs.ruby-lang.org/issues/836)
- [x] Add `StringScanner#captures` [Feature #836](https://bugs.ruby-lang.org/issues/836)
- [x] Add `StringScanner#values_at`  [Feature #836](https://bugs.ruby-lang.org/issues/836)

#### BasicSocket

- [ ] `BasicSocket#read_nonblock` and `BasicSocket#write_nonblock` no longer sets the O_NONBLOCK file description flag as side effect (on Linux only) [Feature #13362](https://bugs.ruby-lang.org/issues/13362)

#### Random

- [x] `Random.raw_seed` renamed to become `Random.urandom`. It is now applicable to non-seeding purposes due to [Bug #9569](https://bugs.ruby-lang.org/issues/9569).

#### Socket::Ifaddr

- [ ] `Socket::Ifaddr#vhid` is added [Feature #13803](https://bugs.ruby-lang.org/issues/13803)

#### ConditionVariable, Queue and SizedQueue reimplemented for speed.

- [x] They no longer subclass `Struct`. [Feature #13552](https://bugs.ruby-lang.org/issues/13552)

#### mathn.rb

- [x] Removed from stdlib. [Feature #10169](https://bugs.ruby-lang.org/issues/10169)

#### Rubygems

- [x] Removed "ubygems.rb" file from stdlib. It's needless since Ruby 1.9.

### C API updates

(Probably not JRuby-relevant.)

### Implementation improvements

(Probably not JRuby-relevant.)

* This might not be a "user visible feature change" but) Hash class's
  hash function is now SipHash13. [Feature #13017](https://bugs.ruby-lang.org/issues/13017)
* SecureRandom now prefers OS-provided sources than OpenSSL. [Bug #9569](https://bugs.ruby-lang.org/issues/9569)
* Mutex rewritten to be smaller and faster [Feature #13517](https://bugs.ruby-lang.org/issues/13517)
* Performance of block passing using block parameters is improved by lazy Proc allocation [Feature #14045](https://bugs.ruby-lang.org/issues/14045)
* Dynamic instrumentation for TracePoint hooks instead of using "trace" instruction to avoid overhead [Feature #14104](https://bugs.ruby-lang.org/issues/14104)

### Miscellaneous changes

- [ ] Print backtrace and error message in reverse order if STDERR is unchanged and a tty. [Feature #8661](https://bugs.ruby-lang.org/issues/8661) [experimental] (Started in #5094)
- [ ] Print error message in bold/underlined text if STDERR is unchanged and a tty. [Feature #14140](https://bugs.ruby-lang.org/issues/14140) [experimental] (Started in #5094)

**Comment by olleolleolle** [comment]
@headius See https://gist.github.com/olleolleolle/10cdc1c94214d77d07e8bd56bd80af20 for a Ruby 2.5 sheet of issues - I believe you know best what to polish away from that, but the leg-work is done.

**Comment by enebo** [comment]
@olleolleolle thanks for all that work.  I added it to the original description!

**Comment by headius** [comment]
@olleolleolle Thank you!

**Comment by headius** [comment]
@olleolleolle I have updated the description by removing any stdlib that we use unmodified (or nearly unmodified). We will still need to toe the line on gemification of those libraries, but I don't consider them major items (a single pass after 2.5 is released to set up all our equivalent default gems will do it).

**Comment by enebo** [comment]
Removed two MRI optimizations about removing GVL from stat-like operations.  We don't have a GVL.

**Comment by headius** [comment]
I believe the new plan, given the long tail of 9.2 development, is to release 9.2 with Ruby 2.5 support.

**Comment by esparta** [comment]
> I have updated the description by removing any stdlib that we use unmodified (or nearly unmodified). We will still need to toe the line on gemification of those libraries, but I don't consider them major items (a single pass after 2.5 is released to set up all our equivalent default gems will do it).

So I was wondering, if stdlib should be updated or not.A few days ago I did a PR for CSV (#5040) as a way to keep up to date. As long as I can tell the gemification is still a WIP, and any (important?) changes are still done on trunk.

**Comment by headius** [comment]
@esparta Yes, it should. We also need to get the tests updated, but I have been waiting on that a bit since a ton of new stuff will fail. The more we get implemented the fewer things we have to work through or exclude from test runs.

The stdlib update process is a bit complicated so I'll take care of that.

**Comment by headius** [comment]
@esparta Oh and thank you for the CSV patch! I will just grab it from a full stdlib update, but that was a very nice way to do a partial update. For future reference, we maintain a fork of the stdlib via https://github.com/jruby/ruby, for example the jruby-ruby_2_3_0 branch. That's where I update stdlib from CRuby, and then as they update I can easily pull in their commits. We do lose the history when importing to JRuby, but they remain visible in the fork repo.

If there's a better way to keep our stdlib up to date, I'm willing to discuss it!

**Comment by headius** [comment]
Ok, I've updated stdlib from 2.5.0 on the ruby-2.5 branch. I did not audit to see if there were new or removed files other than those that broke the sync_ruby script, so help is welcome there.

I prepared the stdlib fork a bit differently this time, starting with a clean branch off the v_2_5_0 tag from CRuby and then cherry-picking the commits I made to previous branches. It's not a great way to manage our patch set.

An easy way for you folks to help us here would be to audit the diffs in our fork and help us get them applied to CRuby in a clean way. I'm a committer, but I have not made this a priority since this system works well enough for me. Reducing the size of the diff would be great, though.

**Comment by headius** [comment]
I have updated the CRuby tests from CRuby 2.5.0. Without some missing parser features (do...ensure is the current blocker) they won't be running, but at least there's no need for PRs to include tests.

You can run individual files using this command line until we get the whole suite to boot properly again:

```
$ jruby test/mri/runner.rb <path to test file> [-v] [-n <name of test>]
```

**Comment by headius** [comment]
@enebo just pushed the updated parser to ruby-2.5, so we can start looking at getting the MRI suite running! Travis build is here: https://travis-ci.org/jruby/jruby/builds/345384059

**Comment by headius** [comment]
We have a completed build with the new MRI tests: https://travis-ci.org/jruby/jruby/jobs/345423405

Less than 300 failures!

**Comment by ChrisBr** [comment]
@headius @kares the initial comment was not up to date anymore so I took the liberty to update it:

### Language changes

- [ ] Top-level constant look-up is removed.  [Feature #11547](https://bugs.ruby-lang.org/issues/11547)
- [ ] rescue/else/ensure are allowed inside do/end blocks.  [Feature #12906](https://bugs.ruby-lang.org/issues/12906)
- [ ] refinements take place in string interpolations.  [Feature #13812](https://bugs.ruby-lang.org/issues/13812)

### Core classes updates (outstanding ones only)

#### Array

- [x]  `Array#append`  [Feature #12746](https://bugs.ruby-lang.org/issues/12746)
- [x]  `Array#prepend`  [Feature #12746](https://bugs.ruby-lang.org/issues/12746)

#### Data

- [x] Now deprecated [Feature #3072](https://bugs.ruby-lang.org/issues/3072)

#### Dir

- [x] `Dir.glob` provides new optional keyword argument, `:base`. [Feature #13056](https://bugs.ruby-lang.org/issues/13056)
- [x] `Dir.children`  [Feature #11302](https://bugs.ruby-lang.org/issues/11302)
- [x] `Dir.each_child`  [Feature #11302](https://bugs.ruby-lang.org/issues/11302)

#### File

- [ ]  :newline option to `File.open` implies text mode now.  [Bug #13350](https://bugs.ruby-lang.org/issues/13350)
- [x]  `File#path` now raises an IOError for files opened with `File::Constants::TMPFILE` option. [Feature #13568](https://bugs.ruby-lang.org/issues/13568)
- [x]  Add `File.lutime`  [Feature #4052](https://bugs.ruby-lang.org/issues/4052)

#### Hash

- [x] `Hash#transform_keys`  [Feature #13583](https://bugs.ruby-lang.org/issues/13583)
- [x] `Hash#transform_keys!`  [Feature #13583](https://bugs.ruby-lang.org/issues/13583)
- [x] `Hash#slice`  [Feature #8499](https://bugs.ruby-lang.org/issues/8499)

#### IO

- [ ] `IO#pread`  [Feature #4532](https://bugs.ruby-lang.org/issues/4532)
- [ ] `IO#pwrite`  [Feature #4532](https://bugs.ruby-lang.org/issues/4532)
- [ ] `IO#copy_stream` tries copy offload with `copy_file_range(2)` [Feature #13867](https://bugs.ruby-lang.org/issues/13867)
- [x] `IO#write` accepts multiple arguments  [Feature #9323](https://bugs.ruby-lang.org/issues/9323)

#### IOError

- [ ] exception message "stream closed" is changed  [Bug #13405](https://bugs.ruby-lang.org/issues/13405)

#### Integer

- [x] `Integer.sqrt`  [Feature #13219](https://bugs.ruby-lang.org/issues/13219)
- [x] `Integer#step` does no longer rescue exceptions when given a step value which cannot be compared with `#>` to 0. [Feature #7688](https://bugs.ruby-lang.org/issues/7688)
- [x] `Integer#{round,floor,ceil,truncate}` now always return an Integer. [Bug #13420](https://bugs.ruby-lang.org/issues/13420)

#### Kernel

- [x] `Kernel#yield_self`  [Feature #6721](https://bugs.ruby-lang.org/issues/6721)
- [x] `Kernel#pp` [Feature #14123](https://bugs.ruby-lang.org/issues/14123)

#### Module

- [x] `Module#attr`, `attr_accessor`, `attr_reader`, `attr_writer` are now public [Feature #14132]
- [x] `Module#define_method`, `alias_method`, `undef_method`, `remove_method` are now public [Feature #14133]

#### Numeric

- [x] Numerical comparison operators (`<` ,`<=`, `>=`, `>`) no longer rescue exceptions of `#coerce`. Return nil in `#coerce` if the coercion is impossible. [Feature #7688](https://bugs.ruby-lang.org/issues/7688)

#### Process

- [ ] Precision of `Process.times` is improved if `getrusage(2)` exists. [Feature #11952](https://bugs.ruby-lang.org/issues/11952)

#### Range

- [x] `Range#initialize` no longer rescue exceptions when comparing begin and
    end with `#<=>` and raise a "bad value for range" ArgumentError
    but instead let the exception from the `#<=>` call go through.
    [Feature #7688](https://bugs.ruby-lang.org/issues/7688)

#### Regexp

- [ ] Update to Onigmo 6.1.1.
- [ ] Support absent operator https://github.com/k-takata/Onigmo/issues/82

#### String

- [x] `String#-@` deduplicates unfrozen strings.  Already-frozen strings remain unchanged for compatibility.  [Feature #13077](https://bugs.ruby-lang.org/issues/13077)
- [x] -"literal" (`String#-@`) optimized to return the same object (same as "literal".freeze in Ruby 2.1+) [Feature #13295](https://bugs.ruby-lang.org/issues/13295)
- [x] `String#{casecmp,casecmp?}` now return nil for non-string arguments instead of raising a TypeError. [Bug #13312](https://bugs.ruby-lang.org/issues/13312)
- [x] `String#delete_prefix` is added to remove prefix [Feature #12694](https://bugs.ruby-lang.org/issues/12694)
- [x] `String#delete_prefix!` is added to remove prefix destructively [Feature #12694](https://bugs.ruby-lang.org/issues/12694)
- [x] `String#delete_suffix` is added to remove suffix [Feature #13665](https://bugs.ruby-lang.org/issues/13665)
- [x] `String#delete_suffix!` is added to remove suffix destructively [Feature #13665](https://bugs.ruby-lang.org/issues/13665)
- [ ] `String#each_grapheme_cluster` and `String#grapheme_clusters` is added to enumerate grapheme clusters [Feature #13780](https://bugs.ruby-lang.org/issues/13780)
- [x] `String#start_with?` supports regexp [Feature #13712](https://bugs.ruby-lang.org/issues/13712)

#### Regexp/String:

- [x] Update Unicode version from 9.0.0 to 10.0.0 [Feature #13685](https://bugs.ruby-lang.org/issues/13685)

#### Thread

- [ ] `Thread#fetch`  [Feature #13009](https://bugs.ruby-lang.org/issues/13009)
- [ ] Description set by `Thread#name=` is now visible on Windows 10.

#### Time

- [ ] `Time#at` receives 3rd argument which specifies the unit of 2nd argument. [Feature #13919](https://bugs.ruby-lang.org/issues/13919)

#### KeyError

- [x] `KeyError#receiver` [Feature #12063](https://bugs.ruby-lang.org/issues/12063)
- [x] `KeyError#key`      [Feature #12063](https://bugs.ruby-lang.org/issues/12063)

### Stdlib updates (outstanding ones only)

#### Bundler

- [ ] Add Bundler to Standard Library. [Feature #12733](https://bugs.ruby-lang.org/issues/12733)

#### Pathname

- [ ] New method: `Pathname#glob` [Feature #7360](https://bugs.ruby-lang.org/issues/7360)

#### Psych

- [ ] Update to Psych 3.0.0.
- [ ] Add `:symbolize_names` option to `Psych.load`, `Psych.safe_load` like `JSON.parse` https://github.com/ruby/psych/pull/333, https://github.com/ruby/psych/pull/337
- [ ] Add `Psych::Handler#event_location` https://github.com/ruby/psych/pull/326
- [ ] Make `frozen string literal = true` https://github.com/ruby/psych/pull/320
- [ ] Preserve time zone offset when deserializing times https://github.com/ruby/psych/pull/316
- [ ] Removed deprecated method aliases for `syck` gem https://github.com/ruby/psych/pull/312

#### RbConfig

- [ ] New constants: `RbConfig::LIMITS` is added to provide the limits of C types. This is available when rbconfig/sizeof is required.

#### Ripper

- [ ] New method: `Ripper#state` is added to tell the state of scanner. [Feature #13686](https://bugs.ruby-lang.org/issues/13686)
- [ ] New constants: `Ripper::EXPR_BEG` and so on for `Ripper#state`.

#### SecureRandom

- [ ] New methods: `SecureRandom.alphanumeric`

#### Set

- [x] Add `Set#to_` as alias to `#inspect` [Feature #13676](https://bugs.ruby-lang.org/issues/13676)
- [x] Add `Set#===` as alias to `#include?` [Feature #13801](https://bugs.ruby-lang.org/issues/13801)
- [x] Add `Set#reset` [Feature #6589](https://bugs.ruby-lang.org/issues/6589)

#### StringIO

- [ ] `StringIO#write` accepts multiple arguments

#### StringScanner

- [ ] Add `StringScanner#size`  [Feature #836](https://bugs.ruby-lang.org/issues/836)
- [ ] Add `StringScanner#captures` [Feature #836](https://bugs.ruby-lang.org/issues/836)
- [ ] Add `StringScanner#values_at`  [Feature #836](https://bugs.ruby-lang.org/issues/836)

### Compatibility issues (excluding feature bug fixes)

#### BasicSocket

- [ ] `BasicSocket#read_nonblock` and `BasicSocket#write_nonblock` no longer sets the O_NONBLOCK file description flag as side effect (on Linux only) [Feature #13362](https://bugs.ruby-lang.org/issues/13362)

#### Random

- [ ] `Random.raw_seed` renamed to become `Random.urandom`. It is now applicable to non-seeding purposes due to [Bug #9569](https://bugs.ruby-lang.org/issues/9569).

#### Socket::Ifaddr

- [ ] `Socket::Ifaddr#vhid` is added [Feature #13803](https://bugs.ruby-lang.org/issues/13803)

#### ConditionVariable, Queue and SizedQueue reimplemented for speed.

- [ ] They no longer subclass `Struct`. [Feature #13552](https://bugs.ruby-lang.org/issues/13552)

### Stdlib compatibility issues (excluding feature bug fixes)

#### mathn.rb

- [ ] Removed from stdlib. [Feature #10169](https://bugs.ruby-lang.org/issues/10169)

#### Rubygems

- [ ] Removed "ubygems.rb" file from stdlib. It's needless since Ruby 1.9.

### C API updates

(Probably not JRuby-relevant.)

### Supported platform changes

#### Drop to support NaCl platform

- [ ] https://bugs.chromium.org/p/chromium/issues/detail?id=239656#c160

### Implementation improvements

(Probably not JRuby-relevant.)

* This might not be a "user visible feature change" but) Hash class's
  hash function is now SipHash13. [Feature #13017](https://bugs.ruby-lang.org/issues/13017)
* SecureRandom now prefers OS-provided sources than OpenSSL. [Bug #9569](https://bugs.ruby-lang.org/issues/9569)
* Mutex rewritten to be smaller and faster [Feature #13517](https://bugs.ruby-lang.org/issues/13517)
* Performance of block passing using block parameters is improved by lazy Proc allocation [Feature #14045](https://bugs.ruby-lang.org/issues/14045)
* Dynamic instrumentation for TracePoint hooks instead of using "trace" instruction to avoid overhead [Feature #14104](https://bugs.ruby-lang.org/issues/14104)

### Miscellaneous changes

- [ ] Print backtrace and error message in reverse order if STDERR is unchanged and a tty. [Feature #8661](https://bugs.ruby-lang.org/issues/8661) [experimental]
- [ ] Print error message in bold/underlined text if STDERR is unchanged and a tty. [Feature #14140](https://bugs.ruby-lang.org/issues/14140) [experimental]
- [ ] configure option --with-ext now mandates its arguments. So for instance if you run ./configure --with-ext=openssl,+ then the openssl library is guaranteed compiled, otherwise the build fails
abnormally. Note however to always add the ",+" at the end of the argument. Otherwise nothing but openssl are built.  [Feature #13302](https://bugs.ruby-lang.org/issues/13302)

**Comment by headius** [comment]
@ChrisBr Thanks! I copied the updated list over and also ticked the box for Thread#fetch.

**Comment by headius** [comment]
We are getting close to being "done enough" with this to merge to master and start tidying things up for release.

For all those interested in helping, the next step is to review the failures in the MRI and ruby/spec suites and either fix them (adding or modifying features as appropriate) or exclude/tag them (for features we won't support or can't support).

A recent build is here. Look at `test:mri:int` and `spec:ruby` results.

**Comment by nomadium** [comment]
@headius There are also a few open PRs with Ruby 2.5 related changes that should help to reduce the number of failures in the tests suites, e.g.: https://github.com/jruby/jruby/pull/5056, https://github.com/jruby/jruby/pull/5061, https://github.com/jruby/jruby/pull/5079, and https://github.com/jruby/jruby/pull/5085.

**Comment by headius** [comment]
@nomadium Thanks! I've merged in the relevant ones. We're getting there!

**Comment by headius** [comment]
The vast bulk of this is now implemented, and I've merged the branch to master. We'll need to green up the builds this week, but we're not far off.

I'll create a new issue for remaining unimplemented 2.5 items.
